### ISSUE TITLE
Support H3 v4.x

### ISSUE THREAD

**Issue by arongergely** [arongergely/qgis-h3-toolkit-plugin#5]
H3 v4.0.0 is to be released in a week (16 August). 

It introduces breaking changes to the API.
See: https://gist.github.com/nrabinowitz/908bd012cd0df4b62d250d0c3b177be2

Either switch to 4.x or add support along along 3.x.
The H3 dependency is not managed by the plugin, so user should be informed about any changes via the install helper dialog

**Comment by arongergely** [comment]
https://github.com/uber/h3/releases/tag/v4.0.0

**Comment by maphew** [comment]
If you get an error like this, you''ve encountered the breaking changes:

```
QGIS version: 3.38.0-Grenoble
QGIS code revision: 37aa6188bc
Qt version: 5.15.13
Python version: 3.12.4
GDAL version: 3.9.0
GEOS version: 3.12.2-CAPI-1.18.2
PROJ version: Rel. 9.4.0, March 1st, 2024
PDAL version: 2.6.3 (git-version: b5523a)
Algorithm started at: 2025-02-04T11:55:34
Algorithm 'Create H3 grid' starting…
Input parameters:
{ 'EXTENT' : '-219287.116900000,1281054.352400000,167805.934600000,1990928.071400000 [EPSG:3578]', 'OUTPUT' : 'TEMPORARY_OUTPUT', 'RESOLUTION' : 1 }

Looking up grid cell indexes...
Traceback (most recent call last):
File "C:\Users/mhwilkie/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\h3_toolkit\processing\algorithms.py", line 254, in processAlgorithm
newSet = h3.polyfill(geoJsonDict, resolution, geo_json_conformant=True)
^^^^^^^^^^^
AttributeError: module 'h3' has no attribute 'polyfill'

Execution failed after 0.10 seconds

Loading resulting layers
Algorithm 'Create H3 grid' finished
```

**Comment by arongergely** [comment]
Yeah. Since h3 v4.0.0 has been superseeded long ago, we could bump the dependency version to 4.x.

I really need to carve out some time to do maintenance on this plugin.

**Comment by maphew** [comment]
I can apply some attention this week or next. Is there anything in particular that would be helpful?

**Comment by arongergely** [comment]
Thank you for your offer. Any contribution is highly appreciated!

I'll review the code andopen issues then get back with ideas tomorrow.

**Comment by arongergely** [comment]
This might just encourage me to also put in some long-overdue time this month

**Comment by arongergely** [comment]
Hey @maphew , if you still feel like giving it a shot you could pick one of these issues:
https://github.com/arongergely/qgis-h3-toolkit-plugin/issues/17
https://github.com/arongergely/qgis-h3-toolkit-plugin/issues/9

I can take care of the h3 v.4 update. Though I may first do some more issue-busting.

**Comment by maphew** [comment]
ok, I'm looking through those issues and sketching out some things. 
I haven't worked on a qgis plugin before, only scripts. Is there a page that talks about how to test a plugin while editing? For example with straight python I can `pip install --editable path/to/code/repo`. Or is it a continual cycle of "zip up changes >>  Qgis: Plugins >> Install from zip >> use plugin" and using the Plugin Reloader Plugin. I'm reviewing https://docs.qgis.org/3.34/en/docs/pyqgis_developer_cookbook/plugins/ide_debugging.html but don't see anything that addresses this (in a way I recognize).

**Comment by arongergely** [comment]
Thanks. yes -  there is a way! 

I set up a symbolic link to my local copy of h3 plugin's dir  to the plugin directory within my qgis profile folder. 
Note: Not to github repo's root dir, but to the `h3_toolkit` dir within)
QGIS then will see the plugin as 'installed' (need to restart QGIS if you havent) .
You may need to still enable the plugin in the plugins window.

The exact way of doing this is particular to the OS you are on, but here are some pointers:

- To find your active profile folder and within it the plugins dir: https://gis.stackexchange.com/questions/274311/what-is-the-default-plugin-folder-location-in-qgis-3
- I'm on linux, so I use `ln -s` to create the symbolic link. 

You can 

Then to reload the plugin into without needing to restart QGIS,  you can use the 
Plugin Reloader plugin: https://plugins.qgis.org/plugins/plugin_reloader

**Comment by arongergely** [comment]
So the link you would want is something like this:
 `<active QGIS profile dir>/python/plugins/h3_toolkit -> /path/to/my/qgis-h3-toolkit-plugin/h3_toolkit`

**Comment by maphew** [comment]
Done, and confirmed working. I'm using a Flatpak Qgis. For searchers, this is the recipe:

```
# grant user filesystem access to the flatpak
flatpak override --user --filesystem=home org.qgis.qgis

# symlink the development directory
ln -s ~/dev/qgis-h3-toolkit-plugin/h3_toolkit ~/.var/app/org.qgis.qgis/data/QGIS/QGIS3/profiles/default/python/plugins/h3-toolkit
```
